### PR TITLE
Replace tmpnam usage

### PR DESCRIPTION
**Pull request by goldsborough**
Fix 
```
/torch_shm_manager#compile-manager.cpp.oc089dac2,gcc-5-glibc-2.23-clang/manager.cpp.o:manager.cpp:function main: 
warning: the use of `tmpnam' is dangerous, better use `mkstemp`
```

@apaszke

### CODE REVIEW

File: torch/lib/libshm/manager.cpp [hunk 1]
```diff
@@ -83,16 +83,19 @@ int main(int argc, char *argv[]) {
   setsid();  // Daemonize the process
 
   std::unique_ptr<ManagerServerSocket> srv_socket;
+  char manager_socket_filename[] = "/tmp/fileXXXXXX";
```
Comment (soumith) [inline torch/lib/libshm/manager.cpp]: isn't this dependent on some env variables, like $TMP, and not always `/tmp`?
also if we have the opportunity, `fileXXXXXX` should be `torch-shm-fileXXXXXX`

Comment (goldsborough) [reply torch/lib/libshm/manager.cpp]: Oh hm, I guess I can `getenv` it

Comment (soumith) [reply torch/lib/libshm/manager.cpp]: ideally, mimic whatever they do inside `std::filesystem::temp_directory_path`

Comment (soumith) [reply torch/lib/libshm/manager.cpp]: https://github.com/llvm-mirror/llvm/blob/0253f7506fa808312119393a633d03cde96cbc8f/lib/Support/Unix/Path.inc#L998-L1037

looks like it checks `"TMPDIR", "TMP", "TEMP", "TEMPDIR"` in that order

**Comment by goldsborough** [comment]
I took my RAII wrapper implementation from `test/cpp/common/suport.h` and beefed it up a bit. @apaszke @soumith

**Review by soumith — APPROVED** [review]

File: test/cpp/api/serialize.cpp [hunk 2]
```diff
@@ -54,7 +54,7 @@ TEST(SerializeTest, BasicToFile) {
 
   auto x = torch::randn({5, 5});
 
-  torch::test::TempFile tempfile;
+  auto tempfile = torch::make_tempfile();
```
Comment (apaszke) [inline test/cpp/api/serialize.cpp]: Do we really want this to appear in the top namespace? Seems like a namespace pollution issue to me.

**Review by apaszke — APPROVED** [review]
One nit, LGTM.

File: test/cpp/api/serialize.cpp (hunk 2 above)
Comment (goldsborough) [reply test/cpp/api/serialize.cpp]: I'll put it into `utils`

**Pull request closed without merging**